### ISSUE TITLE
Adding subtitle text right/left padding

### ISSUE THREAD

**Issue by axionnneeeel** [google/ExoPlayer#7516]
Hello.

Is there any possibility to add padding before and after subtitle text? 

I succeeded changing the appearance using CaptionStyleCompat but can't find anything about adding padding to left or right.

**Comment by icbaker** [comment]
I'm assuming you're using the ExoPlayer UI components (if you're using a custom way to display subtitles then you can presumably just change that directly).

This isn't possible using CaptionStyleCompat.

It is possible by mutating the Cue objects after they're parsed and before they're displayed. There's more information about doing that by subclassing `PlayerView` here: https://github.com/google/ExoPlayer/issues/7243#issuecomment-614651130

When you change the Cues you'll want to overwrite `position`, `positionAnchor` and `size`. If you want to add 10% padding on each side (percentages wrt viewport width) you'd set:
```
// Position the left-hand side of the cue box 10% in from the left edge of the screen:
position = 0.1f;
positionAnchor = ANCHOR_TYPE_START;
// Make the cue box 80% wide, i.e. 10% padding on the right hand side:
size = 0.8f;
```

You can see more info about what all the Cue fields mean in the javadoc:
https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/text/Cue.html

**Comment by nigelmegitt** [comment]
In [IMSC Text Profile](https://www.w3.org/TR/ttml-imsc1.1/#text-linePadding) and [EBU-TT-D](https://tech.ebu.ch/publications/tech3380) profile TTML subtitle files this padding can be specified using the `linePadding` attribute in the `urn:ebu:tt:style` namespace (often assigned the prefix `ebutts:`) - it would be a good improvement to ExoPlayer to support this; it is used by the BBC in its EBU-TT-D subtitles, as specified in the [BBC subtitle guidelines - Background size](http://bbc.github.io/subtitle-guidelines/#Background-size).

In summary, the semantic is that the background area of each line in a `p` is extended in the inline direction at both ends by the amount specified by `linePadding`, which must be expressed in `c` units; those units translate into an absolute size relative to the subtitle rendering area.

**Comment by icbaker** [comment]
Thanks for flagging @nigelmegitt - I filed a separate feature request ticket to track that: https://github.com/google/ExoPlayer/issues/7521

@axionnneeeel As @nigelmegitt indicates, you might also be able to achieve the padding effect by changing your source subtitles - if you control them. You can't use TTML's linePadding because we don't currently support it, but you could use whatever positioning controls your subtitle format offers to achieve a similar effect. If you find that you're able to do this, but ExoPlayer doesn't handle them correctly, please let us know and we can investigate.

**Comment by axionnneeeel** [comment]
Unfortunately, i can't control the source subtitles. So, probably the only thing i can do it the thing you mentioned before.

**Comment by axionnneeeel** [comment]
Hello @icbaker .

I tried your suggestion but i think I got a little misunderstood. 

My windowColor is transparent so i don't want to edit cue box wide, i use a color only for backgroundColor. I just want to add some padding between text and backgroundColor edges. I think that the solution you mentioned is to add padding right/left for the window, not the background.

I also attached a image with the case I am talking about.

![case](https://user-images.githubusercontent.com/19507102/85039050-fc53a000-b18f-11ea-8633-2a82034f35b1.jpg)

**Comment by icbaker** [comment]
Ah right, I was giving instructions to pad the cue box away from the edge of the screen rather than pad the cue text away from the edge of the background box.

I don't think there's a way to achieve what you want. The background color is configured as part of the `TextPaint` that's passed to the `StaticLayout` here:
https://github.com/google/ExoPlayer/blob/6026b919a143ad870cfad3a9d84a260452ebc43b/library/ui/src/main/java/com/google/android/exoplayer2/ui/SubtitlePainter.java#L403 

AFAICT there's no way to tell Android to draw the background with more padding left & right of the text.

If you know your cues are never going to wrap, you could really hackily add non-breaking spaces to the beginning and end of each line by mutating the Cue in the way I described above (changing `Cue.text` instead of size & position). You'd need to be careful to handle any line-breaks in the cue text, and also use `SpannableStringBuilder` to ensure you preserve styling. I'm not sure this is a good idea though, and it will likely break in certain edge cases (and will certainly break if the cues wrap, which they might if your user specifies a large font size).

**Comment by axionnneeeel** [comment]
I think now i have the answer.

Thanks @icbaker .
